### DISCUSSION TITLE
Hajk 3.8.0-alpha.1

### DISCUSSION THREAD

**Discussion by jacobwod** [hajkmap/Hajk#886] [General]
## What's Changed
This is an alpha-release where we ask you to **primarily test the Edit plugin**, as it's seen some substantial changes. 

If you have some time, please test the other major changes in this release. Take a look at the list of **already closed issues labeled with 3.8**: https://github.com/hajkmap/Hajk/milestone/9?closed=1.

In addition, there have been numerous improvements and fixes since 3.7, check out the full changelog here: https://github.com/hajkmap/Hajk/compare/v3.7.0...v3.8.0-alpha.1. 

## How to test
It's super-easy, but here's a quick recap:
1. Download the zip file and unpack
2. Open a terminal window and go to the unpacked directory
3. Install deps: `npm install`
4. Run: `npm start`
5. Open a browser at `http://localhost:3002` for Client UI
6. Open a browser at `http://localhost:3002/admin` for Admin UI
7. Setup your own WFS-T services in Admin UI, add to map and test the Edit plugin! 💥 💯 🚀 
8. Report back in the discussions thread! 🥇 👍 

<hr /><em>This discussion was created from the release <a href='https://github.com/hajkmap/Hajk/releases/tag/v3.8.0-alpha.1'>Hajk 3.8.0-alpha.1</a>.</em>

**Comment by eri-blan** [comment]
Started testing and encountered some problems. It is tested using an Oracle table.

QGIS Server-related:
SERVICE=WFS has to be added to the url in order for the table of attributes to show up. However, when starting the editing tool in the client, an error comes up complaining about it since the service-parameter including "?" is added twice. A workaround is to first enter the url with the SERVICE-parameter, do the configurations, save, remove the SERVICE-parameter and save again. 

In the client:
When starting to write something into a text-field, the geometry on the map disappears.
DateTime attributes for instance are not shown in the editing form.

I´ve only tested briefly, but will do some more tests during the week.

**Reply by jacobwod** [reply]
Thanks @eri-blan, you're absolutely right. 

@hitomin21, can you fix the handling of `?` vs `&`? It's an easy fix, I've done something like this recently in:https://github.com/hajkmap/Hajk/blob/7de36580d340a9f781e966545c6ab73aad1574b0/new-client/src/utils/ConfigMapper.js#L47-L56

The other issues that @eri-blan found need to be fixed too, before we can merge the PR @hitomin21.

**Reply by hitomin21** [reply]
I can fix it but I have no environment where I can test the qgis server related bugs

**Reply by hitomin21** [reply]
@eri-blan : I hope I have solved the issue with double service search parameters but I cannot reproduce that the geometry disappears. I tested on Edge and Chrome on Windows, what did you test on and with what type of config of the tool?

**Reply by hitomin21** [reply]
@eri-blan  Not sure if I misunderstood, does the time part of the selection not show up for you for datetime? It looks like below for me in Firefox (note that it is a datetime and not a timestamp as the label suggests) and similar in Edge

![bild](https://user-images.githubusercontent.com/22234673/139062824-de4da173-ccc8-4ef7-8530-18ba4a8d5f4a.png)

**Reply by eri-blan** [reply]
Maybe there is something odd due to the setup (QGIS Server/Oracle). I´ll try to provide some screenshots and requests/responses if this helps, perhaps you can see something erroneous. It is tested on Edge and Chrome.

Here is an example test setup in admin.
![image](https://user-images.githubusercontent.com/92728323/139199380-72bf9f31-7a79-4f20-bc51-41c5176c8ceb.png)

What is saved to the layers config: 
`{
      "id": "d85ae6a9-f030-2258-fb6f-ad0f3b8a7c4e",
      "caption": "Redigering test",
      "url": "https://hajk.orebro.se/api/v1/proxy/qgisserver/maps/Test/Redigering_test2",
      "uri": "",
      "layers": [
        "TEST"
      ],
      "projection": "EPSG:3009",
      "editableFields": [
        {
          "index": 1,
          "name": "OK_ID",
          "alias": "OK_ID",
          "description": "",
          "dataType": "integer",
          "textType": null,
          "values": null,
          "hidden": false
        },
        {
          "index": 2,
          "name": "TEXT1",
          "alias": "TEXT1",
          "description": "",
          "dataType": "string",
          "textType": "fritext",
          "values": null,
          "hidden": false
        },
        {
          "index": 3,
          "name": "TEXT2",
          "alias": "TEXT2",
          "description": "",
          "dataType": "string",
          "textType": "fritext",
          "values": null,
          "hidden": false
        },
        {
          "index": 4,
          "name": "DATUM",
          "alias": "Date",
          "description": "",
          "dataType": "dateTime",
          "textType": null,
          "values": null,
          "hidden": false
        }
      ],
      "nonEditableFields": [
        {
          "index": 0,
          "name": "geometry",
          "alias": "geometry",
          "description": "",
          "dataType": "PolygonPropertyType",
          "textType": null,
          "values": null,
          "hidden": false
        }
      ],
      "editPoint": true,
      "editMultiPoint": false,
      "editPolygon": true,
      "editMultiPolygon": false,
      "editLine": true,
      "editMultiLine": false
    }`

Part of the getFeature response when choosing a layer to edit (one of the existing features):
`<gml:featureMember>
 <qgs:TEST gml:id="TEST.3">
  <gml:boundedBy>
   <gml:Envelope srsName="EPSG:3009">
    <gml:lowerCorner>164937.9252262 6574169.69159103</gml:lowerCorner>
    <gml:upperCorner>165925.92457282 6574873.43541507</gml:upperCorner>
   </gml:Envelope>
  </gml:boundedBy>
  <qgs:geometry>
   <Polygon xmlns="http://www.opengis.net/gml" srsName="EPSG:3009">
    <exterior xmlns="http://www.opengis.net/gml">
     <LinearRing xmlns="http://www.opengis.net/gml">
      <posList xmlns="http://www.opengis.net/gml" srsDimension="2">164937.9252262 6574596.60536897 165013.1507822 6574241.54074463 165925.92457282 6574169.69159103 165374.23345102 6574873.43541507 164937.9252262 6574596.60536897</posList>
     </LinearRing>
    </exterior>
   </Polygon>
  </qgs:geometry>
  <qgs:OK_ID>3</qgs:OK_ID>
  <qgs:TEXT1 xsi:nil="true"></qgs:TEXT1>
  <qgs:TEXT2 xsi:nil="true"></qgs:TEXT2>
  <qgs:DATUM>2021-10-25T00:00:00.000</qgs:DATUM>
 </qgs:TEST>
</gml:featureMember>`

The resulting form when adding an object:
![image](https://user-images.githubusercontent.com/92728323/139202572-5dd606be-4163-4ce6-a387-fc2ab652582a.png)

Elements not showing up are empty:
`<div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"><span></span></div>`
Seems to be the same for integer and decimal.

Let me know if you want me to send some other data/screenshot @hitomin21. I´ll try to do some debugging as well if i find the time.

**Reply by hitomin21** [reply]
Sorry, I've been sick for a couple of days. I believe the problem is that postgis sends its datatypes as int and not integer and date-time instead of dateTime.

We need to fix the conversion from dataType to textType in getValueMarkup in AttributeEditor in admin. 

@eri-blan: Do you think you could send the data type names generated with your setup (QGIS Server/Oracle, sorry have nowhere to actually test this) for the datatypes below:

date (not datetime), real numbers and booleans

I can fix the code after that.

**Reply by jacobwod** [reply]
Hi @hitomin21, I've seen some pushes to the branch related to your PR (https://github.com/hajkmap/Hajk/pull/884), but I don't think that's all, right? Let me know when you've addressed the issues mentioned in this thread and you want me to prepare another alpha release. I think it's a good way to get as many testers as possible, so we'll continue with this method, but first let me know when you're done and want another alpha.

**Reply by hitomin21** [reply]
It is basically all for now, I have had problems reproducing the issues above and I was sick most of last week. I will make one more push after I get the response regarding Oracle and focus on the reported issues here. I'll write once I think we can generate a new alpha!

**Reply by jacobwod** [reply]
No worries, just let us know whenever you're ready.

**Reply by eri-blan** [reply]
Hi @hitomin21! 
The data type name for real seems to become decimal. 
For date, both date and time is stored in the same data type. There is no date only data type, so date becomes dateTime, even if truncated in the database.
There is also no boolean data type, usually what is used is number or something with the size of 1.

**Reply by hitomin21** [reply]
Pushed a fix that

- adds support for the datatype names from Oracle 
- adds the texttype boolean to integer, which turns them into a checkbox like a normal boolean
- adds the texttype datum to dateTime, which lets the user only specify a date without a time

**Reply by jacobwod** [reply]
Well done @hitomin21, I'll take a look next week! Is there anything else that's left to fix from the first test round, or is that all? If so, let me know and I'll prepare an alpha 2 release next week. 

If @eri-blan has the possibility, it'll be great to test directly from @hitomin21's [feature branch](https://github.com/hajkmap/Hajk/tree/feature/847-support-for-common-datatypes). Let me know if you do the testing locally.

**Reply by hitomin21** [reply]
@eri-blan : Can you test with your setup and see if it is possible to configure correctly with your datatypes?

**Reply by hitomin21** [reply]
@jacobwod I believe everything should be fixed besides the issue that @eri-blan wrote below which I cannot reproduce. Perhaps it was related with the dataypes not matching but it would be great if it works now.

"When starting to write something into a text-field, the geometry on the map disappears."

Once @eri-blan gives the go that the datatypes are written correct, I think we can generate a new alpha for it.

**Comment by eri-blan** [comment]
The coordinates sent to the server seems to be in the map's projection, although the layer´s projection is given in the request, if these differ.

Sending a datetime value to the server seems to result in an Oracle ORA-01841 error (using the data type date in Oracle, haven´t checked the data type timestamp yet since we don´t use it), could be because the default date format is set to YYYY-MM-DD in our database (Oracle). Removing the time property from the request seems to work in getting it through in our case. I'll try and look into it further. I´ve only tested it with already inserted objects from outside Hajk.


[object object] is being sent to the server as a value when using the Flytta(Move) function. I think this occurs when there is no value set in the database for the given attribute. Also, using the Ändra(Alter?) function results in [object object] in the edit form where no value is set in the database.

**Reply by hitomin21** [reply]
I could not reproduce the [object object] but found other issues with empty values retrieved from the database. The fix now is to remove all fields with null (and empty string for flerval) before handling the feature to be modified.

@eri-blan : Can you test if you can get the [object object] again?

**Comment by jacobwod** [comment]
> The coordinates sent to the server seems to be in the map's projection, although the layer´s projection is given in the request, if these differ.

This seems to be a general problem when sending requests to QGIS, please see my last comment in this thread: https://github.com/hajkmap/Hajk/issues/882#issuecomment-949639339

Looks like we will need to rewrite some parts so that all requests are already sent with the layer's SRS, that would solve it, don't you think?

**Comment by jacobwod** [comment]
Probably not our case, but @eri-blan, can you ensure that the requests are being sent to WFS 1.1? I ran across some threads where they indicate that picking the higher version of WFS service would make WFS do the coordinate transformation server-side.

**Reply by eri-blan** [reply]
Yes, the request says v1.1.0

**Reply by jacobwod** [reply]
@eri-blan Please see my last comment in #882, I think it's related. Seems like the WMS project settings also control which SRSes are available in the WFS.
